pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1751,6 +1756,102 @@ func (l *LibvirtDomainManager) GetFilesystems() ([]v1.VirtualMachineInstanceFile
 	return fsList, nil
 }
 
+func (l *LibvirtDomainManager) GetSEVInfo() (*v1.SEVPlatformInfo, error) {
+	sevNodeParameters, err := l.virConn.GetSEVInfo()
+	if err != nil {
+		log.Log.Reason(err).Error("Getting SEV platform info failed")
+		return nil, err
+	}
+
+	return &v1.SEVPlatformInfo{
+		PDH:       sevNodeParameters.PDH,
+		CertChain: sevNodeParameters.CertChain,
+	}, nil
+}
+
+func (l *LibvirtDomainManager) GetLaunchMeasurement(vmi *v1.VirtualMachineInstance) (*v1.SEVMeasurementInfo, error) {
+	domName := api.VMINamespaceKeyFunc(vmi)
+	dom, err := l.virConn.LookupDomainByName(domName)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error(failedGetDomain)
+		return nil, err
+	}
+	defer dom.Free()
+
+	const flags = uint32(0)
+	domainLaunchSecurityParameters, err := dom.GetLaunchSecurityInfo(flags)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Getting launch security info failed")
+		return nil, err
+	}
+
+	sevMeasurementInfo := v1.SEVMeasurementInfo{}
+	if domainLaunchSecurityParameters.SEVMeasurementSet {
+		sevMeasurementInfo.Measurement = domainLaunchSecurityParameters.SEVMeasurement
+	}
+	if domainLaunchSecurityParameters.SEVAPIMajorSet {
+		sevMeasurementInfo.APIMajor = domainLaunchSecurityParameters.SEVAPIMajor
+	}
+	if domainLaunchSecurityParameters.SEVAPIMinorSet {
+		sevMeasurementInfo.APIMinor = domainLaunchSecurityParameters.SEVAPIMinor
+	}
+	if domainLaunchSecurityParameters.SEVBuildIDSet {
+		sevMeasurementInfo.BuildID = domainLaunchSecurityParameters.SEVBuildID
+	}
+	if domainLaunchSecurityParameters.SEVPolicySet {
+		sevMeasurementInfo.Policy = domainLaunchSecurityParameters.SEVPolicy
+	}
+
+	loader := l.efiEnvironment.EFICode(false, true) // no secureBoot, with sev
+	f, err := os.Open(loader)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Errorf("Error opening loader binary %s", loader)
+		return nil, err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		log.Log.Object(vmi).Reason(err).Errorf("Error reading loader binary %s", loader)
+		return nil, err
+	}
+
+	sevMeasurementInfo.LoaderSHA = fmt.Sprintf("%x", h.Sum(nil))
```

Comment:
> So IIUC, above (domainLaunchSecurityParameters.SEVMeasurement) is the state measurement from SEV hardware.

Correct. The measurement includes the initial memory state of the guest.

> Here we add another measurement that is the loader's 256-SHA.

This is not exactly 'another measurement'. This is the SHA digest of the OVMF binary `/usr/share/OVMF/OVMF_CODE.cc.fd` (which is included in the initial memory measuremnet mentioned above). The VM user will need this digest to calculate the measurement locally and then compare with the value returned from SEV hardware.

> So this will be the place to add more custom measurements in the future, right?

Well, in case when the VM is run with external kernel/initrd, the user will also require the appropriate digests of all the compnents: kernel, initrd, cmdline. But this is currently not implemented in this PR. The support of running an SEV VM with external kernel is still in TODO and will be added in a follow-up.

Reply:
Thanks for the clarifications!